File: app/src/main/java/com/example/core/interop/InteropClient.kt
package com.example.core.interop

import android.content.Context
import android.content.pm.PackageManager
import android.database.Cursor
import android.util.Log

/**
 * Reads ChronosFlow's [InteropProvider]. Every call degrades to an empty list when ChronosFlow
 * isn't installed, hasn't granted us access, or otherwise fails — DevTime always works standalone.
 */
class InteropClient(private val context: Context) {

    /**
     * True if ChronosFlow is installed on this device (and visible to us via the manifest
     * `<queries>` entry). Use this to gate any "shared with ChronosFlow" UI — when it's false the
     * app simply runs solo. Cheap enough to call on demand.
     */
    fun isPeerInstalled(): Boolean = try {
        context.packageManager.getPackageInfo(InteropContract.CHRONOSFLOW, 0)
        true
    } catch (e: PackageManager.NameNotFoundException) {
        false
    }

    /**
     * True if ChronosFlow is installed AND exposes a trusted interop provider we're allowed to read
     * (installed + correct authority owner + pinned signing cert). This is the precondition every
     * fetch already checks internally; expose it so callers/UI can branch without issuing a query.
     */
    fun isPeerShareAvailable(): Boolean = isPeerInstalled() && peerProviderTrusted()

    data class RemoteTask(
        val externalId: String,
        val title: String,
        val dueAt: Long?,
        val timezone: String?,
    )

    data class RemoteEvent(
        val externalId: String,
        val title: String,
        val startAt: Long,
        val endAt: Long,
        val timezone: String?,
        val isAllDay: Boolean,
    )

    fun fetchPeerTasks(): List<RemoteTask> = query(InteropContract.PATH_TASKS) { c ->
        RemoteTask(
            externalId = c.string("external_id") ?: return@query null,
            title = c.string("title").orEmpty(),
            dueAt = c.longOrNull("due_at"),
            timezone = c.string("timezone"),
        )
    }

    fun fetchPeerEvents(): List<RemoteEvent> = query(InteropContract.PATH_EVENTS) { c ->
        RemoteEvent(
            externalId = c.string("external_id") ?: return@query null,
            title = c.string("title").orEmpty(),
            startAt = c.longOrNull("start_at") ?: return@query null,
            endAt = c.longOrNull("end_at") ?: 0L,
            timezone = c.string("timezone"),
            isAllDay = (c.longOrNull("is_all_day") ?: 0L) != 0L,
        )
    }

    private fun <T> query(path: String, map: (Cursor) -> T?): List<T> {
        if (!peerProviderTrusted()) return emptyList()
        return try {
            context.contentResolver.query(InteropContract.peerUri(path), null, null, null, null)
                ?.use { c ->
                    buildList {
                        while (c.moveToNext()) map(c)?.let(::add)
                    }
                } ?: emptyList()
        } catch (e: SecurityException) {
            Log.w(TAG, "Interop: peer denied access to /$path: ${e.message}")
            emptyList()
        } catch (e: Exception) {
            // Peer not installed / provider missing / transient failure — stay standalone.
            Log.d(TAG, "Interop: /$path unavailable: ${e.message}")
            emptyList()
        }
    }

    /**
     * Confirms the app that actually owns ChronosFlow's authority is the real, pinned ChronosFlow
     * before we query it — so an app squatting `com.chronosflow.share` can't feed us forged data.
     */
    private fun peerProviderTrusted(): Boolean {
        val info = try {
            context.packageManager.resolveContentProvider(InteropContract.PEER_AUTHORITY, 0)
        } catch (e: Exception) {
            null
        } ?: return false
        if (info.packageName != InteropContract.CHRONOSFLOW) {
            Log.w(TAG, "Interop: ${InteropContract.PEER_AUTHORITY} is owned by ${info.packageName}, not ChronosFlow — ignoring.")
            return false
        }
        return PeerVerifier.isTrusted(context, info.packageName)
    }

    private fun Cursor.string(name: String): String? {
        val i = getColumnIndex(name)
        return if (i < 0 || isNull(i)) null else getString(i)
    }

    private fun Cursor.longOrNull(name: String): Long? {
        val i = getColumnIndex(name)
        return if (i < 0 || isNull(i)) null else getLong(i)
    }

    private companion object {
        const val TAG = "InteropClient"
    }
}

File: ios/MeridianApp/Interop/InteropClient.swift
// InteropClient.swift
// Meridian — iOS 27 / Swift 6
//
// Reads ChronosFlow's interop snapshot from the SHARED App Group container. This is
// the iOS replacement for Android's `InteropClient`, which queried the peer's
// ContentProvider cursors. Every call degrades to an empty result when ChronosFlow
// isn't installed, hasn't published a snapshot, the App Group is unavailable, or the
// payload is corrupt/untrusted — Meridian always works standalone.
//
// There is no live IPC on iOS: "the peer" is simply whatever snapshot file the peer
// last wrote into the container. `isPeerInstalled()` is therefore approximated by
// "has the peer ever published a (non-empty) snapshot here?" — see the doc on that
// method for why this is the correct behavioural analogue.

import Foundation
import os

/// Reads the peer (ChronosFlow) snapshot out of the App Group container.
///
/// `Sendable`: it holds only immutable configuration and touches the filesystem,
/// so it is safe to call from any actor. The sync manager drives it on the main actor.
/// `@unchecked` because the stored `FileManager` is not `Sendable`, but it is process-global
/// and thread-safe for the read-only operations used here.
struct InteropClient: @unchecked Sendable {

    private let log = Logger(subsystem: "com.example.meridian", category: "Interop")

    /// File manager is process-global and thread-safe for the operations we use.
    private let fileManager: FileManager

    init(fileManager: FileManager = .default) {
        self.fileManager = fileManager
    }

    // MARK: - Availability (Android: isPeerInstalled / isPeerShareAvailable)

    /// URL of the peer's snapshot file inside the shared container, or `nil` when the
    /// App Group itself is unavailable (entitlement missing).
    private var peerSnapshotURL: URL? {
        PeerVerifier.sharedContainerURL()?
            .appendingPathComponent(InteropContract.SnapshotFile.peer, isDirectory: false)
    }

    /// `true` when the peer has published a snapshot we can see.
    ///
    /// Android's `isPeerInstalled()` asked the package manager whether ChronosFlow
    /// was installed. iOS apps cannot enumerate other installed apps, but the only
    /// observable consequence of "peer installed" that mattered for sync was "the
    /// peer has data for us to read", which on iOS means "the peer wrote a snapshot
    /// into our shared container". A peer that is installed but has shared nothing
    /// still writes an (empty) snapshot via its exporter, so file presence is the
    /// faithful analogue.
    func isPeerInstalled() -> Bool {
        guard let url = peerSnapshotURL else { return false }
        return fileManager.fileExists(atPath: url.path)
    }

    /// `true` when ChronosFlow has published a snapshot AND that snapshot is trusted
    /// (same-team App Group provenance). Mirrors Android `isPeerShareAvailable()`:
    /// the precondition every fetch checks internally, exposed so the sync manager
    /// can branch (notably for reminder ownership) without re-reading.
    func isPeerShareAvailable() -> Bool {
        PeerVerifier.isPeerShareAvailable(peerSnapshotExists: isPeerInstalled())
            && readSnapshot() != nil
    }

    // MARK: - Fetch (Android: fetchPeerTasks / fetchPeerEvents)

    /// The peer's shareable tasks, or `[]` on any failure. (Android `fetchPeerTasks`.)
    /// Undated tasks are passed through here; the sync manager skips them (parity:
    /// Android filtered `dueAt == null` in `syncFromPeer`).
    func fetchPeerTasks() -> [InteropTask] {
        readSnapshot()?.tasks ?? []
    }

    /// The peer's shareable calendar events, or `[]` on any failure. (Android
    /// `fetchPeerEvents`.)
    func fetchPeerEvents() -> [InteropEvent] {
        readSnapshot()?.events ?? []
    }

    // MARK: - Snapshot read + trust

    /// Reads, decodes and trust-checks the peer snapshot. Returns `nil` (→ empty
    /// fetches, standalone behaviour) on any failure, exactly like Android's `query`
    /// swallowing `SecurityException` / missing-provider exceptions.
    func readSnapshot() -> InteropSnapshot? {
        guard let url = peerSnapshotURL else {
            log.debug("Interop: shared container unavailable — staying standalone.")
            return nil
        }
        guard fileManager.fileExists(atPath: url.path) else {
            log.debug("Interop: peer has not published a snapshot.")
            return nil
        }
        let data: Data
        do {
            data = try Data(contentsOf: url, options: [.mappedIfSafe])
        } catch {
            log.debug("Interop: failed to read peer snapshot: \(error.localizedDescription, privacy: .public)")
            return nil
        }
        let snapshot: InteropSnapshot
        do {
            snapshot = try InteropCoders.decoder.decode(InteropSnapshot.self, from: data)
        } catch {
            log.debug("Interop: peer snapshot undecodable (schema drift?): \(error.localizedDescription, privacy: .public)")
            return nil
        }
        // Android: confirm the owner of the authority is the real, pinned peer before
        // trusting its rows. iOS: confirm same-team App Group provenance + author id.
        guard PeerVerifier.isPeerTrusted(snapshot: snapshot) else {
            log.warning("Interop: peer snapshot failed trust check — ignoring.")
            return nil
        }
        return snapshot
    }
}
